commit-message: add file: tidb-storage.md (#3071)
pingcap/docs: tidb-storage.md

---
title: Storage of TiDB database
summary: Understand the storage layer of a TiDB database

> category: introduction

Reviewer: ```suggestion
```

![Raft in TiDB](/media/tidb-storage-1.png)

To summarize, TiKV can store data on disk quickly via single machine RocksDB, and replicate data to multiple machines via Raft to prevent single machine failure. Instead of writing RocksDB directly, the data is written via the Raft interface. With the implementation of Raft,TiKV becomes a distributed Key-Value storage, and a few machines can automatically complete the replication through the native Raft protocol, which makes it insensitive to the service.

## Region

First, for ease of understanding, in this section, it is assumed that there is only one replica of all data. As mentioned earlier, TiKV can be thought of as a large, ordered KV Map, so data will be distributed across multiple machines in order to achieve a horizontal expansion of storage.
For a KV system, there are two typical solutions for distributing data on multiple machines:

> * Hash: Hash by Key, according to the Hash Value Selection of the corresponding storage node

Reviewer: ```suggestion
* Hash: create Hash by Key and select the corresponding storage node according to the Hash value.
```

> * Range: Divides a Range by Key, where a contiguous Key is stored on a storage node

Reviewer: ```suggestion
* Range: Divide ranges by Key, where a segment of serial Key is stored on a node.
```

> TiKV chose the second way to divide the whole Key-Value space into a series of consecutive Key segments, each segment is called a Region, and will try to keep the data in each Region within a certain size, currently the default is 96 mb in TiKV. Each Region can be described with [StartKey, EndKey] such a left-closed and right-open interval.

Reviewer: ```suggestion
TiKV chooses the second solution that divides the whole Key-Value space into a series of consecutive Key segments. Each segment is called a Region. There is a size limit for each Region to store data (the default value is 96 MB and the size can be configured. Each Region can be described with `[StartKey, EndKey)` such a left-inclusive and right-exclusive interval.
```

![Region in TiDB](/media/tidb-storage-2.png)

> Note that the Region here has nothing to do with the table in SQL. The discussion here still does not involve SQL, only related to KV. After dividing the data into Regions, TiKV will do two important things:

Reviewer: ```suggestion
Note that the Region here has nothing to do with the table in SQL. In this document, please forget about SQL and focus on KV for now. After dividing data into Regions, TiKV will perform two important tasks:
```

> * The data is distributed across all nodes in the cluster in terms of Region, and a similar number of regions is ensured for each node as far as possible

Reviewer: ```suggestion
* Distributing data to all nodes in the cluster and use Region as the basic unit. Ensure that the number of Regions on each node is roughly similar.
```

> * Replication and membership management of Raft based on Region

Reviewer: ```suggestion
* Performing Raft replication and membership management in Region.
```

> These two points are important:

Reviewer: ```suggestion
These two tasks are very important and I'll go through them one by one.
```

```
Key1 -> Value
Key2 -> Value
……
KeyN -> Value
```

With MVCC, the key arrangement of TiKV is like this:

```
Key1_Version3 -> Value
Key1_Version2 -> Value
Key1_Version1 -> Value
……
Key2_Version4 -> Value
Key2_Version3 -> Value
Key2_Version2 -> Value
Key2_Version1 -> Value
……
KeyN_Version2 -> Value
KeyN_Version1 -> Value
……
```

Note that for multiple versions of the same Key, the version with the larger version number will be placed in the front, and the version with the smaller version number will be placed in the back (see the Key-Value section, Keys are arranged in order), so that when the user obtaining Value through a Key + Version, the Key of MVCC can be constructed from Key and Version, which is Key_Version. Then you can directly locate the first position greater than or equal to this Key_Version through RocksDB's SeekPrefix(Key_Version) API.

## Distributed ACID transaction

> TiKV's transaction uses the transaction model used by Google in BigTable: [Percolator](https://research.google.com/pubs/pub36726.html), TiKV is implemented according to this paper, and a lot of optimizations have been made. See the transaction overview for a detailed introduction.

Reviewer: ```suggestion
Transaction of TiKV adopts the model used by Google in BigTable: [Percolator](https://research.google.com/pubs/pub36726.html). TiKV is implemented inspired by this paper, with a lot of optimizations. See [transaction overview](/transaction-overview.md) for details.
```

+ The Key-Value pair in the Map is ordered according to Keys' binary order, which means you can Seek the position of a particular Key and then call the Next method to get the Key-Value pairs larger than this Key in incremental order.

Note that the KV storage model for TiKV described in this document has nothing to do with tables of SQL. This document does not discuss any concepts related to SQL and only focuses on how to implement a high-performance, high-reliability, distributed Key-Value storage such as TiKV.

## Local storage (RocksDB)

For any persistent storage engine, data is eventually saved on disk, and TiKV is no exception. TiKV does not write data directly on the disk, but stores data in RocksDB, which is responsible for the data storage. The reason is that it costs a lot to develop a standalone storage engine, especially a high-performance standalone engine that requires careful optimization.

RocksDB is an excellent standalone storage engine open-sourced by Facebook. This engine can meet various requirements of TiKV for a single engine. Here, you can simply consider RocksDB as a single persistent Key-Value Map.

## Raft protocol

> What's more, the implementation of TiKV faces a more difficult thing: how to secure data safety in case a single machine fails?

Reviewer: ```suggestion
What's more, the implementation of TiKV faces a more difficult thing: to secure data safety in case a single machine fails.
```

![Raft in TiDB](/media/tidb-storage-1.png)

In summary, TiKV can quickly store data on disk via the standalone machine RocksDB, and replicate data to multiple machines via Raft in case of machine failure. Data is written through the interface of Raft instead of to RocksDB. With the implementation of Raft, TiKV becomes a distributed Key-Value storage. Even with a few machine failures, TiKV can automatically complete replicas by virtue of the native Raft protocol, which does not impact the application.

## Region

To make it easy to understand, let's assume that all data only has one replica. As mentioned earlier, TiKV can be regarded as a large, orderly KV Map, so data is distributed across multiple machines in order to achieve horizontal scalability. For a KV system, there are two typical solutions to distributing data across multiple machines:

* Hash: Create Hash by Key and select the corresponding storage node according to the Hash value.
* Range: Divide ranges by Key, where a segment of serial Key is stored on a node.

> TiKV chooses the second solution that divides the whole Key-Value space into a series of consecutive Key segments. Each segment is called a Region. There is a size limit for each Region to store data (the default value is 96 MB and the size can be configured). Each Region can be described by `[StartKey, EndKey)`, a left-inclusive and right-exclusive interval.

Reviewer 2: ```suggestion
TiKV chooses the second solution that divides the whole Key-Value space into a series of consecutive Key segments. Each segment is called a Region. There is a size limit for each Region to store data (the default value is 96 MB and the size can be configured). Each Region can be described by `[StartKey, EndKey)`, a left-closed and right-open interval.
```

See https://en.wikipedia.org/wiki/Interval_(mathematics)#Terminology

![Region in TiDB](/media/tidb-storage-2.png)

Note that the Region here has nothing to do with the table in SQL. In this document, forget about SQL and focus on KV for now. After dividing data into Regions, TiKV will perform two important tasks:

> * Distributing data to all nodes in the cluster and use Region as the basic unit. Ensure that the number of Regions on each node is roughly similar.

Reviewer 2: ```suggestion
* Distributing data to all nodes in the cluster and use Region as the basic unit. Try its best to ensure that the number of Regions on each node is roughly similar.
```

* Performing Raft replication and membership management in Region.

These two tasks are very important and will be introduced one by one.

* First, data is divided into many Regions according to Key, and the data for each Region is stored on only one node (ignoring multiple replicas). The TiDB system has a PD component that is responsible for spreading Regions as evenly as possible across all nodes in the cluster. In this way, on one hand, the storage capacity is scaled horizontally (Regions on the other nodes are automatically scheduled to the newly added node); on the other hand, load balancing is achieved (the situation where one node has a lot of data while the others have little will not occur).

    At the same time, in order to ensure that the upper client can access the needed data, there is a component (PD) in the system to record the distribution of Regions on the node, that is, the exact Region of a Key and the node of that Region placed through any Key.

> * For the second task, TiKV replicates data in Regions, which means that data in one Region will have multiple replicas with the name “Replica”. Multiple replicas of a Region are stored on different nodes to form a Raft Group, which is kept consistent through the Raft algorithm.

Reviewer 2: ```suggestion
* For the second task, TiKV replicates data in Regions, which means that data in one Region will have multiple replicas with the name “Replica”. Multiple Replicas of a Region are stored on different nodes to form a Raft Group, which is kept consistent through the Raft algorithm.
```

>     One of the replicas serves as the Leader of the Group and other as the Follower. By default, all reads and writes are processed through the Leader, where reads are done and write are replicated to followers. The following diagram shows the whole picture about Region and Raft group.

Reviewer 2: ```suggestion
    One of the Replicas serves as the Leader of the Group and other as the Follower. By default, all reads and writes are processed through the Leader, where reads are done and write are replicated to followers. The following diagram shows the whole picture about Region and Raft group.
```

Note that the KV storage model for TiKV described in this document has nothing to do with tables in SQL. This document does not discuss any concepts related to SQL and focuses on how to implement a high-performance, high-reliability, distributed Key-Value storage such as TiKV.

## Local storage (RocksDB)

For any persistent storage engine, data is eventually saved on disk, and TiKV is no exception. TiKV does not write data directly on the disk, but stores data in RocksDB, which is responsible for the data storage. The reason is that it costs a lot to develop a standalone storage engine, especially a high-performance standalone engine that requires careful optimization,
RocksDB is an excellent standalone storage engine open-sourced by Facebook. This engine can meet various requirements of TiKV for a single engine. Here, you can simply consider RocksDB as a single persistent Key-Value Map.

## Raft protocol

What's more, the implementation of TiKV faces a more difficult thing: how to secure data safety in case a single machine fails?

> A simple way is to replicate data to multiple machines, so that even if one machine fails, the replicas on other machines are still available. To put it difficult, you need a data replication scheme that is reliable, efficient, and able to handle the situation of a failed replica. All of these are made possible by the Raft algorithm.

Reviewer: ```suggestion
A simple way is to replicate data to multiple machines, so that even if one machine fails, the replicas on other machines are still available. In other words, you need a data replication scheme that is reliable, efficient, and able to handle the situation of a failed replica. All of these are made possible by the Raft algorithm.
```
